feat: support csv in `fga tuple delete`

bensooraj opened the pull request:
<!-- Thanks for opening a PR! Here are some quick tips:
If this is your first time contributing, [read our Contributing Guidelines](https://github.com/openfga/.github/blob/main/CONTRIBUTING.md) to learn how to create an acceptable PR for this repo.
By submitting a PR to this repository, you agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md)

If your PR is under active development, please submit it as a "draft". Once it's ready, open it up for review.
-->

<!-- Provide a brief summary of the changes -->

## Description
* `fga tuple delete --file=(*.json|*.yaml|*csv)`
* The tuple delete command now supports reading values of a `*.csv` file as well

## References
<!-- Provide a list of any applicable references here (GitHub Issue, [OpenFGA RFC](https://github.com/openfga/rfcs), other PRs, etc..) -->
* This PR is a copy of [feat: support csv in tuple delete](https://github.com/openfga/cli/pull/260)
* Documentation: [doc: fga tuple delete --file=(csv|json|yaml) now supports CSV](https://github.com/openfga/openfga.dev/pull/1004)
* Relevant issues:
   1. [[Feature] Support csv in tuple delete #477](https://github.com/openfga/cli/issues/477)
   2. [tuple delete should support CSV files #256](https://github.com/openfga/cli/issues/256)

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

Siddhant-K-code commented:
👋🏻 @bensooraj, I saw you ticked `I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]"


Could you link the docs PR too? As, it would be super helpful if we can have docs change too along with the product side change. Let me know, if you need any help with that.

**For context:** We have our docs [here](https://github.com/openfga/openfga.dev)

Siddhant-K-code commented on internal/tuplefile/read.go:
```diff
@@ -32,3 +32,16 @@ func ReadTupleFile(fileName string) ([]client.ClientTupleKey, error) {
 	return tuples, nil
 }
+
+func TupleKeysToTupleKeysWithoutCondition(tks []client.ClientTupleKey) []client.ClientTupleKeyWithoutCondition {
```
**Suggestion (non-blocking):**  

Add documentation something like following:

   ```go
   // TupleKeysToTupleKeysWithoutCondition converts ClientTupleKey to ClientTupleKeyWithoutCondition,
   // stripping out condition-related fields.
   ```

bensooraj replied:
Thank you for the suggestion, I have incorporated this.

Siddhant-K-code commented on cmd/tuple/delete_test.go:
Consider adding test cases for:
   - Empty JSON/YAML files
   - Malformed JSON/YAML files
   - Edge cases with special characters in CSV files

bensooraj replied:
Hey @Siddhant-K-code,

A few clarifying questions around your suggestion:
1. Empty JSON/YAML files don't throw a parsing error. `[]` and `{}` are valid JSON as well. Do the following test cases make sense?
```go
....
{
	name:          "empty json file should throw a warning",
	file:          "testdata/tuples_empty.json",
	expectedError: "found empty json file: no action taken",
},
{
	name:          "empty yaml file should throw a warning",
	file:          "testdata/tuples_empty.yaml",
	expectedError: "found empty yaml file: no action taken",
},
....
```
2. By malformed JSON/YAML files, do you mean invalid format or incomplete tuples (example given below)?
```json
[
    {
        // missing user
        "relation": "owner",
        "object": "folder:product"
    },
    {
        "user": "folder:product",
        // missing relation
        "object": "folder:product-2021"
    },
    {
        "user": "user:beth",
        "relation": "viewer",
        // missing object
    }
]
```
3. Do we have a documentation of what characters are allowed for `user`, `relation`, `object` and/or `condition`. I could only find the following in the documentation,
> In OpenFGA, relations can only have alphanumeric characters, underscores and hyphens. We recommend using underscore (_) to separate words and removing prepositions. E.g.: "can create a document" can become "can_create_document" or "create_document" if you are into brevity. - source: [Get Started with Modeling](https://openfga.dev/docs/modeling/getting-started#03-list-relations-for-those-types)

I do see we have some basic (regex) checks available at `github.com/openfga/openfga/pkg/tuple`. I can implement those checks (`func Is*() {...}`) at the CLI and continue to depend on the `openfga` server for anything beyond the regex validations. IMO, that should be a separate PR.

Siddhant-K-code replied:
> Empty JSON/YAML files don't throw a parsing error. [] and {} are valid JSON as well. Do the following test cases make sense?

Yeah, that looks good!

> By malformed JSON/YAML files, do you mean invalid format or incomplete tuples (examples given below)?


I think, it should be already covered by server side, so, i don't think, we need to add that.


> I see we have some basic (regex) checks available at github.com/openfga/openfga/pkg/tuple. I can implement those checks (func Is*() {...}) at the CLI and continue to depend on the openfga server for anything beyond the regex validations. IMO, that should be a separate PR.

Yeah, you are right, it should be consistent and should be just handled by server side only

bensooraj replied:
Thanks, @Siddhant-K-code. I have incorporated the cases for empty json and yaml files.

Siddhant-K-code reviewed (commented):
Great PR! The changes improve code organization and add solid test coverage. A few minor suggestions for future improvements (suggested below)

These are minor suggestions - the PR is solid and ready to merge as is. Thanks for the improvements!

bensooraj commented:
> 👋🏻 @bensooraj, I saw you ticked `I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]"`
> 
> Could you link the docs PR too? As, it would be super helpful if we can have docs change too along with the product side change. Let me know, if you need any help with that.
> 
> **For context:** We have our docs [here](https://github.com/openfga/openfga.dev)

I have added documentation at [openfga.dev#1004](https://github.com/openfga/openfga.dev/pull/1004)

Siddhant-K-code commented:
@bensooraj can you please fix the lint errs

Siddhant-K-code commented on .gitignore:
```diff
@@ -33,4 +33,7 @@ dist/
\ No newline at end of file
+/tests/fixtures/identifiers
+
+# IDE configs
+.vscode
```
You can move this to above at L27

bensooraj replied:
Done.

bensooraj commented:
> @bensooraj can you please fix the lint errs

I thought I did, my bad. I have fixed those as well.

Siddhant-K-code reviewed (approved):
Changes looks good to me.

Will also someone from @openfga/dx team for a final review & then we can merge

dyeam0 commented:
@bensooraj Thanks for the contribution and, @Siddhant-K-code, thanks for the review!

We will assign someone in our planning meeting this week to pick this one up for a final review.

dyeam0 commented:
Hi @bensooraj, we discussed this PR during our planning meeting. We've put it in the queue for someone to pick up. They will have a full review for you next week. Thanks!

rhamzeh reviewed (approved):
Thanks @bensooraj for the contribution and @Siddhant-K-code for the help reviewing!

rhamzeh commented:
Needs #489 to be merged first so that the lint issues pass, then we can merge this
